File: webshared/src/main/java/com/taskadapter/web/uiapi/NewMappingBuilder.java
package com.taskadapter.web.uiapi;

import com.taskadapter.connector.definition.AvailableFields;
import com.taskadapter.connector.definition.FieldMapping;
import com.taskadapter.connector.definition.NewMappings;
import com.taskadapter.model.GTaskDescriptor;

public class NewMappingBuilder {
    static NewMappings createNewMappings(AvailableFields m1, AvailableFields m2) {
        final NewMappings res = new NewMappings();
        for (GTaskDescriptor.FIELD field : GTaskDescriptor.FIELD.values()) {
            if (field == GTaskDescriptor.FIELD.ID || field == GTaskDescriptor.FIELD.REMOTE_ID) {
                continue;
            }

            if (!m1.isFieldSupported(field) || !m2.isFieldSupported(field)) {
                continue;
            }

            /* Don't create mappings here. New mappings will be generated in the fix-up phase.
             */
            final boolean selectByDefault = m1.isSelectedByDefault(field)
                    && m2.isSelectedByDefault(field);

            res.put(new FieldMapping(field, m1.getDefaultValue(field), m2
                    .getDefaultValue(field), selectByDefault));
        }

        if (m2.isFieldSupported(GTaskDescriptor.FIELD.REMOTE_ID)) {
            res.put(new FieldMapping(GTaskDescriptor.FIELD.REMOTE_ID, null, 
                    m2.getDefaultValue(GTaskDescriptor.FIELD.REMOTE_ID), 
                    m2.isSelectedByDefault(GTaskDescriptor.FIELD.REMOTE_ID)));
        }

        if (m1.isFieldSupported(GTaskDescriptor.FIELD.REMOTE_ID)) {
            res.put(new FieldMapping(GTaskDescriptor.FIELD.REMOTE_ID, 
                    m1.getDefaultValue(GTaskDescriptor.FIELD.REMOTE_ID), null, 
                    m1.isSelectedByDefault(GTaskDescriptor.FIELD.REMOTE_ID)));
        }

        return res;

    }
}

File: webui/src/main/scala/com/taskadapter/web/uiapi/NewMappingBuilder.scala
package com.taskadapter.web.uiapi

import java.util

import com.taskadapter.connector.Field
import com.taskadapter.connector.definition.FieldMapping

import scala.collection.JavaConverters._

object NewMappingBuilder {
  val DEFAULT_VALUE_FOR_EMPTY_VALUES = ""

  /**
    * try to match list of fields for connector 1 with the list for connector 2.
    */
  def suggestedFieldMappingsForNewConfig(config1: UIConnectorConfig, config2: UIConnectorConfig): util.List[FieldMapping] = {
    // TODO TA3 restore remote ids
    /*
            if (m2.isFieldSupported(GTaskDescriptor.FIELD.REMOTE_ID)) {
                res.put(new FieldMapping(GTaskDescriptor.FIELD.REMOTE_ID, null,
                        m2.getDefaultValue(GTaskDescriptor.FIELD.REMOTE_ID),
                        m2.isSelectedByDefault(GTaskDescriptor.FIELD.REMOTE_ID),
                        DEFAULT_VALUE_FOR_EMPTY_VALUES));
            }

            if (m1.isFieldSupported(GTaskDescriptor.FIELD.REMOTE_ID)) {
                res.put(new FieldMapping(GTaskDescriptor.FIELD.REMOTE_ID,
                        m1.getDefaultValue(GTaskDescriptor.FIELD.REMOTE_ID), null,
                        m1.isSelectedByDefault(GTaskDescriptor.FIELD.REMOTE_ID),
                        DEFAULT_VALUE_FOR_EMPTY_VALUES));
            }
    */

    val selectByDefault = true

    val connector1Combinations = config1.getSuggestedCombinations
    val connector2Combinations = config2.getSuggestedCombinations

    val result = scala.collection.mutable.ListBuffer[FieldMapping]()
    connector1Combinations.values.foreach{ standardField =>
      val field1 : Field = connector1Combinations.find(i => i._2 == standardField).map(e => e._1).getOrElse(Field(""))
      val field2 : Field = connector2Combinations.find(i => i._2 == standardField).map(e => e._1).getOrElse(Field(""))

      result += FieldMapping(field1, field2, selectByDefault, "")
    }


    result.asJava
  }
}
